### PR TITLE
[WIP] VR Support

### PR DESCRIPTION
**Pull request by jojolepro**
## Description

This adds basic virtual reality support. It was originally developed by magnonellie and then updated by me.

Know issues: 
* The VR headset shows only the gl clear color.
* The remotes are not showing, because the repository for the models is empty.

For testing, make a local copy of this branch (xr) in a folder called "amethystxr". Then, clone this repository next to it https://github.com/jojolepro/amethyst-openvr and run the example.

This PR was only tested on steamvr "htc vive" devices.

I would appreciate if someone with more experience with rendering could check out why the headset isn't showing the scene.

## Additions

- Basic OpenVR support

## PR Checklist

By placing an x in the boxes I certify that I have:

- [ ] Ran `cargo test --all` locally if this modified any rs files.
- [ ] Ran `cargo +stable fmt --all` locally if this modified any rs files.
- [ ] Updated the content of the book if this PR would make the book outdated.
- [ ] Added a changelog entry if this will impact users, or modified more than 5 lines of Rust that wasn't a doc comment.
- [ ] Added unit tests for new APIs if any were added in this PR.
- [X] Acknowledged that by making this pull request I release this code under an MIT/Apache 2.0 dual licensing scheme.

### CODE REVIEW

File: amethyst_xr/src/lib.rs [hunk 1]
```diff
@@ -0,0 +1,142 @@
...
+    fn get_removed_trackers(&mut self) -> Option<Vec<u32>>;
+
+    fn get_tracker_position(&mut self, index: u32) -> TrackerPositionData;
+
+    fn get_area(&mut self) -> Vec<[f32; 3]>;
+    fn get_hidden_area_mesh(&mut self) -> Vec<[f32; 3]>;
+
+    fn get_tracker_models(&mut self, index: u32) -> TrackerModelLoadStatus;
+
+    fn get_gl_target_info(&mut self, near: f32, far: f32) -> Vec<XRTargetInfo>;
```
Comment (Frizi) [inline amethyst_xr/src/lib.rs]: This is obviously very gl specific. While this is probably not a concern right now, we will have to handle multiple backends here.

**Comment by dotellie** [comment]
Oh hey, this is my code! Just as a heads up sort of thing, there's a lot of WIP code in here that I never finished and that won't work too well in the future. I would **highly** recommend to wait until rendy is integrated before merging any XR stuff. I will probably come back to fully implement this once rendy has landed though if you guys don't manage to do it before me. 😉

(Also, I should note that I did at one point manage to correctly render the controllers in the headset, albeit with low framerate and colorspace issues)

**Comment by jojolepro** [comment]
Currently I'm not making any progress on this, because
1) rendy is coming
2) I have literally no idea why its not binding the render targets correctly
3) How do you slow down time in real life? :thinking:

**Comment by Frizi** [comment]
I am pretty positive that this will be much easier to work on once rendy is in place. I think that there is no point on working on it with current renderer.

**Comment by ConnorBP** [comment]
Has rendy been added yet? I'm curious if anyone is looking into implementing this since a performance-oriented rust game engine is in my mind the perfect engine for something like vr where performance is so valuable.

**Comment by fhaynes** [comment]
@ConnorBP Should be about 1-2 more weeks. You can follow the progress on the `rendy` branch. =)

**Comment by mralve** [comment]
OOhhh I can't wait to boot up my Vive to play with this! Awesome pr!

**Comment by TimonPost** [comment]
Is anyone going to do a rebase? :)

**Comment by jojolepro** [comment]
We can't rebase this PR. It currently stays open only as a reference for the next person that will try to integrate VR.

**Comment by jojolepro** [comment]
The branch will stay in my fork, but I'm closing the PR due to inactivity. If you read this and want to help with VR/XR support, feel free to port the code from this PR into a new one.

**Comment by jakbyte** [comment]
@dotellie Are you going to implement this now that rendy has been integrated?

**Comment by dotellie** [comment]
@jakbyte Probably not unfortunately, at least not for quite a while. Other things have come up so I doubt I'll be able to come around to this anytime soon. :slightly_frowning_face:

**Pull request closed without merging**